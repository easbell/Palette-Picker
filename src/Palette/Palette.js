import React, {Component} from 'react';
import Color from '../Color/Color';
import PaletteForm from '../PaletteForm/PaletteForm';

export class Palette extends Component {
  constructor() {
    super()
    this.state = {
      colors: [],
      showForm: false
    }
  }

  componentDidMount() {
    const colors = this.setColors()
    this.setState({colors})
  }

  setColors = () => {
    let colors = [0, 0, 0, 0, 0]
    let randomColor;
    return colors.map(color => {
      randomColor = this.createColors()
      if(!colors.includes(randomColor)) {
        return randomColor
      } else {
        return this.createColors()
      }
    })
  }

  createColors = () => {
    let digits = '0123456789ABCDEF'
    let hexCode = '#'
    for( let i = 0; i < 6; i++) {
      hexCode += digits[(Math.floor(Math.random() * 16))]
    }
    return hexCode
  }

  renderColors = () => {
    const { colors } = this.state
    return colors.map(color => {
      return <Color key={color} color={color} savePalette={this.savePalette}/>
    })
  }

  savePalette = () => {
    console.log(this.state.colors)
    this.setState({ showForm: true })
  }

  render() {
    const { showForm, colors } = this.state;
    return(
      <div>
        <button onClick={this.savePalette}>Save Palete</button>
        {this.renderColors()}
        {showForm &&
          <PaletteForm {...colors}/>
        }
      </div>
    )
  }
}

export default Palette