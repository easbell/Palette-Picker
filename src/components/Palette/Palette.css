.palette {
  display: flex;
  width: 100%;
  height: 80vh;
  align-items: stretch;
  align-content: stretch;
  justify-content: stretch;
}

.controls {
  background: #ea4c88;
  color: #fff;
  box-shadow: 0 3px 5px rgba(0,0,0,0.20);
  padding: 10px 15px;
  font-size: 13px;
  font-weight: bold;
  letter-spacing: .02em;
  line-height: 17px;
  vertical-align: middle;
  cursor: pointer;
  border: none;
  text-decoration: none;
  border-radius: 4px;
  margin: 20px;
}

.controls:focus {
  outline: none;
}

.controls:hover {
  background-color: #ea4c88ad;
  transition: all .2s ease;
}