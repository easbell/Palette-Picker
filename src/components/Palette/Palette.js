import React, {Component} from 'react';
import Color from '../Color/Color';
import PaletteForm from '../../containers/PaletteForm/PaletteForm';
import './Palette.css';

export class Palette extends Component {
  constructor() {
    super()
    this.state = {
      color1: {color: '', locked: false},
      color2: {color: '', locked: false},
      color3: {color: '', locked: false},
      color4: {color: '', locked: false},
      color5: {color: '', locked: false},
      showForm: false
    }
  }

  componentDidMount() {
    this.setColors()
  }
  
  setColors = () => {
    Object.keys(this.state).forEach(color => {
      if(color.includes('color') && this.state[color]['locked'] === false) {
        this.setState({ [color]: { color: this.createColors(), locked: false }})
      }
    })
  }

  createColors = () => {
    let digits = '0123456789ABCDEF'
    let hexCode = '#'
    for( let i = 0; i < 6; i++) {
      hexCode += digits[(Math.floor(Math.random() * 16))]
    }
    return hexCode
  }

  renderColors = () => {
    const stateKeys = Object.keys(this.state)
    const colors = stateKeys.splice(0, stateKeys.length - 1)
    return colors.map((color, i) => {
      const hexCode = this.state[color].color
      return <Color key={i} lockColor={this.lockColor} color={hexCode} savePalette={this.savePalette}/>
    })
  }

  lockColor = (color) => {
    const stateKeys = Object.keys(this.state)
    const stateColors = stateKeys.splice(0, stateKeys.length - 1)
    stateColors.forEach(stateColor => {
      if(this.state[stateColor].color === color) {
        this.setState({ [stateColor]: { color: color, locked: 
        !this.state[stateColor].locked }})
      }
      }
    )
  }

  savePalette = (bool) => {
    this.setState({ showForm: bool })
  }

  render() {
    const { showForm } = this.state;
    return(
      <div>
        <div className='palette'>
          {this.renderColors()}
        </div>
        <button onClick={() => this.savePalette(true)} className='save controls'>Save Palette</button>
        <button onClick={this.setColors} className='controls'>New Palette</button>
        {showForm &&
          <PaletteForm savePalette={this.savePalette} colors={{...this.state}}/>
        }
      </div>
    )
  }
}

export default Palette