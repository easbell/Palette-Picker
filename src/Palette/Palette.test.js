import React from 'react';
import { Palette } from './Palette';
import { shallow } from 'enzyme';

describe('Palette', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Palette />)
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('should set state with updated colors when setColors is called', () => {
    // wrapper.instance().componentDidMount();
    expect(wrapper.instance().setState).toHaveBeenCalled();
  });

  it('should set state when save palette is called', () => {
    wrapper.find('.save').simulate('click', true);
    expect(wrapper.state('showForm')).toBe(true);
  });
});
