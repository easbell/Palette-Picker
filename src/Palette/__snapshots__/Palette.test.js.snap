// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palette should match snapshot 1`] = `
<div>
  <button
    className="save"
    onClick={[Function]}
  >
    Save Palette
  </button>
  <button
    onClick={[Function]}
  >
    New Palette
  </button>
  <div
    className="palette"
  >
    <Color
      color="#FFC47E"
      key="0"
      lockColor={[Function]}
      savePalette={[Function]}
    />
    <Color
      color="#1063FA"
      key="1"
      lockColor={[Function]}
      savePalette={[Function]}
    />
    <Color
      color="#9E0DD4"
      key="2"
      lockColor={[Function]}
      savePalette={[Function]}
    />
    <Color
      color="#7752E5"
      key="3"
      lockColor={[Function]}
      savePalette={[Function]}
    />
    <Color
      color="#09EB7E"
      key="4"
      lockColor={[Function]}
      savePalette={[Function]}
    />
  </div>
</div>
`;
